### ISSUE TITLE
x/tools/gopls: clarify supported operating systems

### ISSUE THREAD

**Issue by findleyr** [golang/go#59981]
We have experienced a fair amount of friction with gopls tests on e.g. plan9, android, illumos, and solaris.  While test failures on these operating systems _occasionally_ turn up real gopls bugs, in my experience failures are more often due to builder problems, resource constraints, and/or bugs in the toolchain. Given that (I assume) few if any of our users are on some of these operating systems, it is almost certain that our resources devoted to these builders results in a net loss for our users.

Unlike other projects on the Go team, gopls is designed for a relatively narrow use-case: development environments. We should not expend resources on ports that are unlikely to serve this use-case for a large number of users. However, we do not specify supported operating systems in our [README](https://github.com/golang/tools/blob/master/gopls/README.md).

We should specify precisely which operating systems we aim to support, and disable builders for others.  At the very least, we will support linux, windows, darwin, and other BSD-like operating systems. I am open to others, but do not want to commit to support if there is no demand.

CC @bcmills @adonovan

**Comment by bcmills** [comment]
I agree that we do not need to explicitly support `android`, although after https://go.dev/cl/488655 I suspect that the [two Android-specific skips in `gopls`](https://cs.opensource.google/search?q=%5C%22android%5C%22%20filepath:gopls&sq=&ss=go%2Fx%2Ftools) might no longer be necessary.

(The only other recent changes that have been needed for Android are also needed to support `go test` on the `gopls` module when run as a dependency of another module, which I think we _should_ aim to support, provided that that module also requires an appropriate version of `x/tools`.)

**Comment by bcmills** [comment]
The problems on `illumos` and `solaris` seem to be due to builder and/or kernel issues. Those have been reported to the port maintainers (as #59379 and #58782 respectively), so per our [porting policy](https://go.dev/wiki/PortingPolicy) I think we can safely ignore `gopls` failures on those builders until those issues are addressed.

**Comment by bcmills** [comment]
As for Plan 9, well, I'm not sure. 😅 The `plan9-arm` builder, at least, seems to do ok with the `gopls` tests most of the time.

We occasionally need to add test skips for lack of symlink support, but in theory we also need those for certain versions and/or configurations of Windows, and those skip checks can call general `testenv` functions that are not specific to Plan 9.

Beyond that we have only two plan9-specific workarounds in place:
- One test skip for #50477, which remains undiagnosed.
- And one [retry loop in `(*Workdir).ReadFile`](https://cs.opensource.google/go/x/tools/+/master:gopls/internal/lsp/fake/workdir.go;l=165-171;drc=6eb432f3b01c1737995d178a7eac935c84ef63c1) to allow `gopls` to read files that are concurrently opened by `go` commands.

To be honest, though, I'm not entirely sure why we don't need an analogous retry loop on Windows. 😅

**Comment by bcmills** [comment]
> We should specify precisely which operating systems we aim to support, and disable builders for others.

That may be tricky. Do we currently have a mechanism to skip only a specific submodule within a repo?
I hesitate to drop all of `x/tools` on a given platform just because `gopls` doesn't support it.

**Comment by bcmills** [comment]
I think mostly what I'm saying is: given that we probably do want to keep testing the rest of `x/tools` on most platforms, and given the “curb cut” effects of many of the fixes (for example: fixing the tests on Android also fixes them when run from the module cache; fixing symlink issues on plan9 also fixes them in certain Windows configurations), the engineering cost of skipping `gopls` tests on the less-supported builders might turn out to be higher than the cost of not skipping them.

**Comment by bsiegert** [comment]
Please don't drop support for the BSDs (FreeBSD, NetBSD, OpenBSD).

**Comment by findleyr** [comment]
@bsiegert we can do that, but can you explain a bit more? e.g. do you think this affects a large number of users? Is it strategically important to support these GOOS?

**Comment by PaluMacil** [comment]
> @bsiegert we can do that, but can you explain a bit more? e.g. do you think this affects a large number of users? Is it strategically important to support these GOOS?

I don't use a BSD, but I know that these are platforms commonly used as everyday desktops besides just servers. While I do not expect someone to sit down and write software on Android, and I have never met someone with a need for Solaris, plan9, etc, I would expect that a lot of BSD users are writing software while sitting at a BSD desktop.

**Comment by findleyr** [comment]
@PaluMacil thanks. Yes, I think there are probably some users on BSD, whereas I expect ~zero users on Android or Plan9. Also, generally speaking, BSDs have been less of a problem for us than others.

**Comment by findleyr** [comment]
@adonovan I see few but nonzero users on 386 at telemetry.go.dev.

We should formally deprecate 386, and follow our normal deprecation procedures (make a warning for one minor release).
